packages/core/src/execution-engine/ssh-clients-manager.ts

```diff
@@ -11,21 +12,54 @@ class SSHClientsConfig {
 	idleTimeout: number = 5 * 60;
 }
 
+type Registration = {
+	client: Client;
+
+	/**
+	 * We keep this timestamp to check if a client hasn't been used in a while,
+	 * and if it needs to be closed.
+	 */
+	lastUsed: Date;
+
+	abortController: AbortController;
+};
+
 @Service()
 export class SSHClientsManager {
-	readonly clients = new Map<string, { client: Client; lastUsed: Date }>();
+	readonly clients = new Map<string, Registration>();
+
+	readonly clientsReversed = new WeakMap<Client, string>();
 
-	constructor(private readonly config: SSHClientsConfig) {
+	private cleanupTimer: NodeJS.Timeout;
+
+	constructor(
+		private readonly config: SSHClientsConfig,
+		private readonly logger: Logger,
+	) {
 		// Close all SSH connections when the process exits
 		process.on('exit', () => this.onShutdown());
 
-		if (process.env.NODE_ENV === 'test') return;
-
 		// Regularly close stale SSH connections
-		setInterval(() => this.cleanupStaleConnections(), 60 * 1000);
+		this.cleanupTimer = setInterval(() => this.cleanupStaleConnections(), 60 * 1000);
+
+		this.logger = logger.scoped('ssh-client');
+	}
+
+	updateLastUsed(client: Client) {
+		const key = this.clientsReversed.get(client);
+
+		if (key) {
+			const registration = this.clients.get(key);
+
+			if (registration) {
+				registration.lastUsed = new Date();
+			}
+		}
 	}
 
-	async getClient(credentials: SSHCredentials): Promise<Client> {
+	async getClient(credentials: SSHCredentials, abortController?: AbortController): Promise<Client> {
```

Comment:
Comes from before but wondering if there's a race condition where we create two SSH connections to the same destination? Or it doesn't matter because if so the orphan connection won't be in the map and won't be found by your new GC method, BUT it will eventually be terminated by the SSH server which will trigger `withCleanupHandler`?

Reply:
That's a really good find. I should fix this. If the server does not close the connection it will stay up forever.